src/v/archival/archiver_manager.cc

```diff
@@ -0,0 +1,1051 @@
+                  if (fut.failed()) {
+                      fsm.process_event(
+                        ev_archiver_failure{.error = fut.get_exception()});
+                  } else {
+                      fsm.process_event(
+                        ev_archiver_started{.archiver = fut.get()});
+                  }
+              };
+            ssx::background = fsm.create_and_start_archiver(new_leadership.term)
+                                .then_wrapped(completion);
+        }
+    };
+
+    struct tr_starting_to_active {
+        void operator()(
+          const ev_archiver_started& e,
+          state_machine_t& fsm,
+          st_starting_async& prev,
+          st_active& next) {
+            vlog(fsm._ctxlog.info, "ntp_archiver started");
+            prev.clear();
+            next.archiver = e.archiver;
+        }
+    };
+
+    struct tr_starting_to_passive {
+        void operator()(
+          const ev_archiver_failure& f,
+          state_machine_t& fsm,
+          st_starting_async& prev,
+          st_passive& next) {
+            // NOTE: it's extremely unlikely for the ntp_archiver.start() to
+            // fail. This method only syncs the FSM and starts the background
+            // fiber. The 'sync' method can throw if it fails to allocate. Same
+            // is true for the `ssx::spawn_with_gate` call. The danger of not
+            // running the archiver is that the broker may run out of disk
+            // space. Because the data is not uploaded to the cloud storage the
+            // max_collectible_offset can't be propagated forward. Because of
+            // that the local data can't be evicted.
+            vlog(
+              fsm._ctxlog.error,
+              "ALERT! ntp_archiver failed to start {}",
+              f.error);
+            prev.clear();
+            if (next.trigger) {
+                next.trigger->set_exception(f.error);
+            }
+        }
+    };
+
+    struct tr_active_to_stopping {
+        template<class Event>
+        void operator()(
+          const Event&,
+          state_machine_t& fsm,
+          st_active& prev,
+          st_stopping_async& next) {
+            vlog(fsm._ctxlog.info, "stopping ntp_archiver");
+            next.reset_units = fsm._part->get_archiver_reset_units();
+            vassert(
+              next.reset_units.has_value(), "Reset units are not acquired");
+            auto completion = [&fsm](ss::future<> fut) {
+                if (fut.failed()) {
+                    vlog(
+                      fsm._ctxlog.error,
+                      "ntp_archiver shutdown failed {}",
+                      fut.get_exception());
+                    fsm.process_event(
+                      ev_archiver_failure{.error = fut.get_exception()});
+                } else {
+                    fsm.process_event(ev_archiver_stopped{});
+                }
+            };
+            ssx::background = fsm.stop_and_dispose_archiver(prev.archiver)
```

Comment:
I'm liking the direction of this change more and more, especially if it means the archiver being responsible for exactly one term